File: app/models/api/content.py
# ruff: noqa: F401
from __future__ import annotations

from datetime import datetime
from enum import StrEnum
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.constants import TWEET_SUGGESTION_MODEL
from app.models.api.content_discussions import ContentDiscussionResponse
from app.models.api.pagination import PaginationMetadata
from app.models.contracts import (
    ContentClassification,
    ContentStatus,
    ContentType,
    SummaryKind,
    SummaryVersion,
)


class ContentSummaryResponse(BaseModel):
    """Summary information for a content item in list view."""

    id: int = Field(..., description="Unique identifier")
    content_type: ContentType = Field(..., description="Type of content (article/podcast/news)")
    url: str = Field(..., description="Canonical URL of the content")
    source_url: str | None = Field(None, description="Original scraped/submitted URL")
    discussion_url: str | None = Field(
        None, description="Discussion URL (tweet, HN thread, etc.) when available"
    )
    title: str | None = Field(None, description="Content title")
    source: str | None = Field(
        None, description="Content source (e.g., substack name, podcast name)"
    )
    platform: str | None = Field(
        None, description="Content platform (e.g., twitter, substack, youtube)"
    )
    status: ContentStatus = Field(..., description="Processing status")
    short_summary: str | None = Field(
        None,
        description=(
            "Short summary for display; for news items this returns the excerpt or first item text"
        ),
    )
    created_at: str = Field(..., description="ISO timestamp when content was created")
    processed_at: str | None = Field(None, description="ISO timestamp when content was processed")
    classification: ContentClassification | None = Field(
        None, description="Content classification (to_read/skip)"
    )
    publication_date: str | None = Field(
        None, description="ISO timestamp of when content was published"
    )
    is_read: bool = Field(False, description="Whether the content has been marked as read")
    is_saved_to_knowledge: bool = Field(
        False,
        description="Whether the content has been saved to the user's knowledge library",
    )
    news_article_url: str | None = Field(
        None, description="Canonical article link for news content"
    )
    news_discussion_url: str | None = Field(
        None, description="Aggregator discussion URL (HN thread, tweet, etc.)"
    )
    news_key_points: list[str] | None = Field(
        None, description="Key points provided for news items"
    )
    news_summary: str | None = Field(None, description="Short overview synthesized for news items")
    user_status: str | None = Field(
        None, description="Per-user content status (e.g., inbox, archived)"
    )
    image_url: str | None = Field(
        None, description="URL of full-size AI-generated image for this content"
    )
    thumbnail_url: str | None = Field(
        None, description="URL of 200px thumbnail image for fast loading in list views"
    )
    primary_topic: str | None = Field(
        None, description="Primary topic extracted from summary topics or platform name"
    )
    top_comment: dict[str, str] | None = Field(
        None, description="First discussion comment {author, text} for preview"
    )
    comment_count: int | None = Field(
        None, description="Discussion comment count from aggregator or discussion fetcher"
    )
    feed_preview: dict[str, Any] | None = Field(
        None, description="Longform artifact feed preview payload"
    )
    artifact_type: str | None = Field(None, description="Longform artifact type")
    preview_bullets: list[str] | None = Field(
        None, description="Longform artifact feed-preview bullets"
    )
    reason_to_read: str | None = Field(
        None, description="Feed-preview reason explaining why the item is worth opening"
    )
    key_takeaway: str | None = Field(
        None, description="Key takeaway to display under long-form list titles"
    )
    saved_source: Literal["knowledge", "x_bookmark"] | None = Field(
        None,
        description="Saved-library source for this content when it appears in saved views",
    )

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "id": 123,
                "content_type": "article",
                "url": "https://example.com/article",
                "title": "Understanding AI in 2025",
                "source": "Tech Blog",
                "platform": "substack",
                "status": "completed",
                "short_summary": "This article explores the latest developments in AI...",
                "created_at": "2025-06-19T10:30:00Z",
                "processed_at": "2025-06-19T10:35:00Z",
                "classification": "to_read",
                "publication_date": "2025-06-18T12:00:00Z",
                "is_read": False,
                "image_url": "/static/images/content/123.png",
                "thumbnail_url": "/static/images/thumbnails/123.png",
                "primary_topic": "AI",
                "top_comment": {"author": "user123", "text": "Great article!"},
                "saved_source": "knowledge",
            }
        }
    )


class ContentListResponse(BaseModel):
    """Response for content list endpoint."""

    contents: list[ContentSummaryResponse] = Field(..., description="List of content items")
    available_dates: list[str] = Field(..., description="List of available dates (YYYY-MM-DD)")
    content_types: list[ContentType] = Field(
        ..., description="Available content types for filtering"
    )
    meta: PaginationMetadata = Field(..., description="Pagination metadata for the response")

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "contents": [
                    {
                        "id": 123,
                        "content_type": "article",
                        "url": "https://example.com/article",
                        "title": "Understanding AI in 2025",
                        "source": "Tech Blog",
                        "platform": "substack",
                        "status": "completed",
                        "short_summary": "This article explores...",
                        "created_at": "2025-06-19T10:30:00Z",
                        "processed_at": "2025-06-19T10:35:00Z",
                        "classification": "to_read",
                    }
                ],
                "available_dates": ["2025-06-19", "2025-06-18"],
                "content_types": ["article", "podcast", "news"],
                "meta": {
                    "next_cursor": "eyJsYXN0X2lkIjoxMjN9",
                    "has_more": True,
                    "page_size": 25,
                    "total": 1,
                },
            }
        }
    )


class NarrationResponse(BaseModel):
    """Unified narration payload for any supported narration target."""

    target_type: Literal["content"] = Field(
        ...,
        description="Narration target family",
    )
    target_id: int = Field(..., description="Target identifier within the target family")
    title: str = Field(..., description="Resolved title for spoken playback")
    narration_text: str = Field(..., description="Plain-text narration script for voice playback")


class DetectedFeed(BaseModel):
    """Detected RSS/Atom feed from content page."""

    url: str = Field(..., description="Feed URL")
    type: str = Field(..., description="Feed type: substack, podcast_rss, or atom")
    title: str | None = Field(None, description="Feed title from link tag")
    format: str = Field("rss", description="Feed format: rss or atom")

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "url": "https://example.substack.com/feed",
                "type": "substack",
                "title": "Example Newsletter",
                "format": "rss",
            }
        }
    )


class SubmissionFeedInitialDownloadResponse(BaseModel):
    """Initial download result for a newly subscribed feed."""

    requested_count: int | None = Field(
        None, description="Number of recent feed items requested for initial download"
    )
    ran: bool | None = Field(None, description="Whether initial download was attempted")
    status: str | None = Field(None, description="Initial download status")
    reason: str | None = Field(None, description="Reason initial download was skipped")
    error: str | None = Field(None, description="Initial download failure reason")
    config_id: int | None = Field(None, description="Feed config used for the download")
    base_limit: int | None = Field(None, description="Original scraper limit")
    target_limit: int | None = Field(None, description="Temporary scraper limit used")
    scraped: int | None = Field(None, description="Number of items scraped")
    saved: int | None = Field(None, description="Number of items saved")
    duplicates: int | None = Field(None, description="Number of duplicate items ignored")
    errors: int | None = Field(None, description="Number of scraper item errors")


class SubmissionFeedSubscriptionResponse(BaseModel):
    """Feed subscription outcome attached to a submission status row."""

    status: str = Field(..., description="Raw feed subscription status")
    feed_url: str | None = Field(None, description="Subscribed feed URL")
    feed_type: str | None = Field(None, description="Subscribed feed type")
    created: bool | None = Field(None, description="Whether a new feed config was created")
    config_id: int | None = Field(None, description="Created feed config identifier")
    initial_download: SubmissionFeedInitialDownloadResponse | None = Field(
        None, description="Initial recent-item download result"
    )


SubmissionKind = Literal["content", "feed_subscription"]
SubmissionOutcome = Literal[
    "queued",
    "processing",
    "completed",
    "failed",
    "skipped",
    "subscribed",
    "already_subscribed",
    "feed_not_found",
    "feed_fetch_failed",
    "feed_subscription_failed",
]


class SubmissionStatusResponse(BaseModel):
    """Status information for a user-submitted content item."""

    id: int = Field(..., description="Unique identifier")
    content_type: ContentType = Field(
        ..., description="Type of content (article/podcast/news/unknown)"
    )
    url: str = Field(..., description="Canonical URL of the content")
    source_url: str | None = Field(None, description="Original submitted URL")
    title: str | None = Field(None, description="Content title (if detected)")
    status: ContentStatus = Field(..., description="Processing status")
    error_message: str | None = Field(None, description="Failure reason when status=failed/skipped")
    created_at: str = Field(..., description="ISO timestamp when content was created")
    processed_at: str | None = Field(None, description="ISO timestamp when content was processed")
    submitted_via: str | None = Field(None, description="Submission channel (share_sheet, etc.)")
    is_self_submission: bool = Field(
        True, description="Whether this content was submitted by the current user"
    )
    submission_kind: SubmissionKind = Field(
        "content", description="Semantic submission kind for user-facing display"
    )
    outcome: SubmissionOutcome = Field(
        "processing", description="Semantic submission outcome for user-facing display"
    )
    detected_feed: DetectedFeed | None = Field(
        None, description="RSS/Atom feed detected while handling a feed subscription request"
    )
    feed_subscription: SubmissionFeedSubscriptionResponse | None = Field(
        None, description="Feed subscription result for Add Feed submissions"
    )


class SubmissionStatusListResponse(BaseModel):
    """Response for user submission status list."""

    submissions: list[SubmissionStatusResponse] = Field(
        ..., description="List of user-submitted items still processing or failed"
    )
    meta: PaginationMetadata = Field(..., description="Pagination metadata for the response")


class PodcastEpisodeSearchResultResponse(BaseModel):
    """Single podcast episode search result."""

    title: str
    episode_url: str
    podcast_title: str | None = None
    source: str | None = None
    snippet: str | None = None
    feed_url: str | None = None
    published_at: str | None = None
    provider: str | None = None
    score: float | None = None


class PodcastEpisodeSearchResponse(BaseModel):
    """Response payload for podcast episode search."""

    results: list[PodcastEpisodeSearchResultResponse] = Field(default_factory=list)


class MixedSearchFeedResultResponse(BaseModel):
    """Validated feed/source result for mixed search."""

    id: str
    title: str
    site_url: str
    feed_url: str
    feed_type: str
    feed_format: str
    description: str | None = None
    rationale: str | None = None
    evidence_url: str | None = None


class MixedSearchResponse(BaseModel):
    """Sectioned mixed search results for the More > Search screen."""

    query: str
    content: list[ContentSummaryResponse] = Field(default_factory=list)
    feeds: list[MixedSearchFeedResultResponse] = Field(default_factory=list)
    podcasts: list[PodcastEpisodeSearchResultResponse] = Field(default_factory=list)


class ContentDetailResponse(BaseModel):
    """Detailed response for a single content item."""

    id: int = Field(..., description="Unique identifier")
    content_type: ContentType = Field(..., description="Type of content (article/podcast/news)")
    url: str = Field(..., description="Canonical URL of the content")
    source_url: str | None = Field(None, description="Original scraped/submitted URL")
    discussion_url: str | None = Field(
        None, description="Discussion URL (tweet, HN thread, etc.) when available"
    )
    title: str | None = Field(None, description="Content title")
    display_title: str = Field(
        ..., description="Display title (prefers summary title over content title)"
    )
    source: str | None = Field(None, description="Content source")
    status: ContentStatus = Field(..., description="Processing status")
    error_message: str | None = Field(None, description="Error message if processing failed")
    retry_count: int = Field(..., description="Number of retry attempts")
    metadata: dict[str, Any] = Field(..., description="Content-specific metadata")
    created_at: str = Field(..., description="ISO timestamp when content was created")
    updated_at: str | None = Field(None, description="ISO timestamp of last update")
    processed_at: str | None = Field(None, description="ISO timestamp when content was processed")
    checked_out_by: str | None = Field(None, description="Worker ID that checked out this content")
    checked_out_at: str | None = Field(
        None, description="ISO timestamp when content was checked out"
    )
    publication_date: str | None = Field(
        None, description="ISO timestamp of when content was published"
    )
    is_read: bool = Field(False, description="Whether the content has been marked as read")
    is_saved_to_knowledge: bool = Field(
        False,
        description="Whether the content has been saved to the user's knowledge library",
    )
    # Additional useful properties from ContentData
    summary: str | None = Field(None, description="Summary text")
    short_summary: str | None = Field(None, description="Short version of summary for list view")
    summary_kind: SummaryKind | None = Field(
        None, description="Summary kind discriminator (e.g., long_interleaved)"
    )
    summary_version: SummaryVersion | None = Field(
        None, description="Summary schema version for the current summary kind"
    )
    structured_summary: dict[str, Any] | None = Field(
        None, description="Structured summary with bullet points and quotes"
    )
    longform_artifact: dict[str, Any] | None = Field(
        None, description="Typed longform artifact envelope"
    )
    feed_preview: dict[str, Any] | None = Field(
        None, description="Longform artifact feed preview payload"
    )
    artifact_type: str | None = Field(None, description="Longform artifact type")
    preview_bullets: list[str] | None = Field(
        None, description="Longform artifact feed-preview bullets"
    )
    reason_to_read: str | None = Field(
        None, description="Feed-preview reason explaining why the item is worth opening"
    )
    bullet_points: list[dict[str, str]] = Field(
        ..., description="Bullet points from structured summary"
    )
    quotes: list[dict[str, str]] = Field(..., description="Quotes from structured summary")
    topics: list[str] = Field(..., description="Topics from structured summary")
    full_markdown: str | None = Field(
        None, description="Full article content formatted as markdown"
    )
    body_available: bool = Field(False, description="Whether canonical body text is available")
    body_kind: str | None = Field(
        None, description="Resolved body kind (article, transcript, or source)"
    )
    body_format: str | None = Field(None, description="Resolved body format (text or markdown)")
    news_article_url: str | None = Field(
        None, description="Canonical article link for news content"
    )
    news_discussion_url: str | None = Field(
        None, description="Aggregator discussion URL (HN thread, tweet, etc.)"
    )
    news_key_points: list[str] | None = Field(
        None, description="Key points provided for news items"
    )
    news_summary: str | None = Field(None, description="Short overview synthesized for news items")
    image_url: str | None = Field(
        None, description="URL of full-size AI-generated image for this content"
    )
    thumbnail_url: str | None = Field(
        None, description="URL of 200px thumbnail image for fast loading"
    )
    detected_feed: DetectedFeed | None = Field(
        None, description="Detected RSS/Atom feed for this content"
    )
    can_subscribe: bool = Field(
        False,
        description="Whether the current user can subscribe to the detected feed",
    )

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "id": 123,
                "content_type": "article",
                "url": "https://example.com/article",
                "title": "Understanding AI in 2025",
                "source": "Tech Blog",
                "status": "completed",
                "error_message": None,
                "retry_count": 0,
                "metadata": {
                    "source": "Tech Blog",
                    "author": "Jane Doe",
                    "publication_date": "2025-06-19T00:00:00Z",
                    "content_type": "html",
                    "word_count": 1500,
                    "summary": {
                        "title": "Understanding AI in 2025",
                        "overview": "This article explores the latest developments...",
                        "bullet_points": [
                            {"text": "AI is transforming industries", "category": "key_finding"}
                        ],
                        "quotes": [{"text": "The future is now", "context": "Jane Doe"}],
                        "topics": ["AI", "Technology", "Future"],
                        "summarization_date": "2025-06-19T10:35:00Z",
                        "classification": "to_read",
                    },
                },
                "created_at": "2025-06-19T10:30:00Z",
                "updated_at": "2025-06-19T10:35:00Z",
                "processed_at": "2025-06-19T10:35:00Z",
                "checked_out_by": None,
                "checked_out_at": None,
                "publication_date": "2025-06-18T12:00:00Z",
                "is_read": False,
                "display_title": "Understanding AI in 2025",
                "summary": "This article explores the latest developments...",
                "short_summary": "This article explores the latest developments...",
                "structured_summary": {
                    "title": "Understanding AI in 2025",
                    "overview": "This article explores the latest developments...",
                    "bullet_points": [
                        {"text": "AI is transforming industries", "category": "key_finding"}
                    ],
                    "quotes": [{"text": "The future is now", "context": "Jane Doe"}],
                    "topics": ["AI", "Technology", "Future"],
                    "summarization_date": "2025-06-19T10:35:00Z",
                    "classification": "to_read",
                },
                "bullet_points": [
                    {"text": "AI is transforming industries", "category": "key_finding"}
                ],
                "quotes": [{"text": "The future is now", "context": "Jane Doe"}],
                "topics": ["AI", "Technology", "Future"],
                "full_markdown": "# Understanding AI in 2025\n\nFull article content...",
                "body_available": True,
                "body_kind": "article",
                "body_format": "text",
                "image_url": "/static/images/content/123.png",
                "thumbnail_url": "/static/images/thumbnails/123.png",
                "can_subscribe": False,
            }
        }
    )


class ContentBodyResponse(BaseModel):
    """Resolved canonical body payload."""

    content_id: int = Field(..., description="Content identifier")
    variant: str = Field(..., description="Canonical body variant")
    kind: str = Field(..., description="Body kind")
    format: str = Field(..., description="Body format")
    text: str = Field(..., description="Full canonical body text")
    updated_at: str | None = Field(None, description="ISO timestamp of the body pointer update")

File: rust/crates/newsly-api/src/content_read/presentation/content.rs
use super::{
    ArtifactFields, ContentClassification, ContentDetailProjection, ContentDetailResponse,
    ContentStatus, ContentSummaryResponse, ContentType, DISPLAY_SUMMARY_KINDS, FromStr,
    LegacyNewsFields, Map, PresentationError, SavedSource, Value, backfill_legacy_news_article,
    clean_optional_str, content_top_comment, detected_feed, extract_key_takeaway,
    extract_short_summary, insert_string_if_missing, is_typed_artifact_detail,
    legacy_content_news_fields, longform_artifact_fields, normalize_summary_contract, object_field,
    parse_summary_kind, parse_summary_version, project_bullet_points, project_quotes,
    project_topics, resolve_content_display_title, resolve_content_image_urls, resolve_content_url,
    runtime_metadata, sanitize_metadata_for_api, string_field, structured_summary,
};

// Normalized source values shared by the two independent wire projections.
struct ContentPresentation {
    row: ContentDetailProjection,
    content_type: ContentType,
    status: ContentStatus,
    metadata: Map<String, Value>,
    url: String,
    display_title: String,
    summary_text: Option<String>,
    news_fields: Option<LegacyNewsFields>,
    discussion_url: Option<String>,
    image_url: Option<String>,
    thumbnail_url: Option<String>,
}

impl ContentPresentation {
    fn new(row: ContentDetailProjection) -> Result<Self, PresentationError> {
        let content_type = ContentType::from_str(&row.content_type)
            .map_err(|_| PresentationError::UnknownContentType(row.content_type.clone()))?;
        let status = ContentStatus::from_str(&row.status)
            .map_err(|_| PresentationError::UnknownContentStatus(row.status.clone()))?;
        let mut metadata = runtime_metadata(&row.content_metadata);
        insert_string_if_missing(&mut metadata, "platform", row.platform.as_deref());
        insert_string_if_missing(&mut metadata, "source", row.source.as_deref());
        backfill_legacy_news_article(&mut metadata, &row, content_type);
        normalize_summary_contract(&mut metadata, content_type);
        let url = resolve_content_url(&row.url, &metadata, content_type)
            .ok_or_else(|| PresentationError::InvalidContentUrl(row.url.clone()))?;
        let display_title = resolve_content_display_title(row.title.as_deref(), &metadata);
        let summary_text = extract_short_summary(object_field(&metadata, "summary"));
        let news_fields = (content_type == ContentType::News
            && longform_artifact_fields(&metadata)
                .longform_artifact
                .is_none())
        .then(|| legacy_content_news_fields(&metadata, &url));
        let discussion_url = news_fields.as_ref().map_or_else(
            || string_field(&metadata, "discussion_url"),
            |fields| fields.discussion_url.clone(),
        );
        let (image_url, thumbnail_url) =
            resolve_content_image_urls(row.id, content_type, &metadata);
        Ok(Self {
            row,
            content_type,
            status,
            metadata,
            url,
            display_title,
            summary_text,
            news_fields,
            discussion_url,
            image_url,
            thumbnail_url,
        })
    }
}

pub(crate) fn present_content_detail(
    row: ContentDetailProjection,
) -> Result<ContentDetailResponse, PresentationError> {
    let content = ContentPresentation::new(row)?;
    let row = content.row;
    let mut metadata = content.metadata;
    let summary_kind = parse_summary_kind(metadata.get("summary_kind"));
    let summary_version = parse_summary_version(metadata.get("summary_version"));
    let artifact = longform_artifact_fields(&metadata);
    let canonical_artifact_detail =
        is_typed_artifact_detail(content.content_type, summary_kind, &artifact);
    let is_legacy_news = content.news_fields.is_some();

    let (structured_summary, bullet_points, quotes, topics) =
        if is_legacy_news || canonical_artifact_detail {
            (None, Vec::new(), Vec::new(), Vec::new())
        } else {
            (
                structured_summary(&metadata, summary_kind),
                project_bullet_points(&metadata, summary_kind, summary_version),
                project_quotes(&metadata, summary_kind, summary_version),
                project_topics(&metadata, summary_kind, summary_version),
            )
        };
    let detected_feed = detected_feed(&metadata);

    let longform_artifact = artifact.longform_artifact.cloned();
    let ArtifactFields {
        mut feed_preview,
        mut artifact_type,
        mut preview_bullets,
        mut reason_to_read,
        ..
    } = artifact;
    if canonical_artifact_detail {
        // The complete envelope owns these values. Keep nullable wire keys for installed clients.
        for key in ["summary", "feed_preview", "selection_trace"] {
            metadata.remove(key);
        }
        feed_preview = None;
        artifact_type = None;
        preview_bullets = None;
        reason_to_read = None;
    }

    Ok(ContentDetailResponse {
        id: row.id,
        content_type: content.content_type,
        url: content.url,
        source_url: row.source_url.or(Some(row.url)),
        discussion_url: content.discussion_url,
        title: row.title,
        display_title: content.display_title,
        source: row.source,
        status: content.status,
        error_message: row.error_message,
        retry_count: row.retry_count,
        metadata: sanitize_metadata_for_api(metadata),
        created_at: row.created_at,
        updated_at: row.updated_at,
        processed_at: row.processed_at,
        checked_out_by: row.checked_out_by,
        checked_out_at: row.checked_out_at,
        publication_date: row.publication_date,
        is_read: row.is_read,
        is_saved_to_knowledge: row.is_saved_to_knowledge,
        summary: content.summary_text.clone(),
        short_summary: content.summary_text.clone(),
        summary_kind,
        summary_version,
        structured_summary,
        longform_artifact,
        feed_preview,
        artifact_type,
        preview_bullets,
        reason_to_read,
        bullet_points,
        quotes,
        topics,
        full_markdown: None,
        body_available: row.body_available,
        body_kind: row.body_available.then(|| {
            if content.content_type == ContentType::Podcast {
                "transcript".to_owned()
            } else {
                "article".to_owned()
            }
        }),
        body_format: row.body_format,
        news_article_url: content
            .news_fields
            .as_ref()
            .and_then(|fields| fields.article_url.clone()),
        news_discussion_url: content
            .news_fields
            .as_ref()
            .and_then(|fields| fields.discussion_url.clone()),
        news_key_points: content
            .news_fields
            .as_ref()
            .and_then(|fields| fields.key_points.clone()),
        news_summary: content.summary_text,
        image_url: content.image_url,
        thumbnail_url: content.thumbnail_url,
        detected_feed,
        can_subscribe: false,
    })
}

pub(crate) fn present_content_summary(
    row: ContentDetailProjection,
    knowledge_saved_at: Option<chrono::DateTime<chrono::Utc>>,
    saved_source_override: Option<SavedSource>,
) -> Result<ContentSummaryResponse, PresentationError> {
    let content = ContentPresentation::new(row)?;
    let row = content.row;
    let metadata = content.metadata;
    let platform = row.platform.clone();
    let kind = parse_summary_kind(metadata.get("summary_kind"));
    let version = parse_summary_version(metadata.get("summary_version"));
    let ArtifactFields {
        feed_preview,
        artifact_type,
        preview_bullets,
        reason_to_read,
        ..
    } = longform_artifact_fields(&metadata);
    let is_legacy_news = content.news_fields.is_some();
    let classification = kind
        .filter(|kind| !is_legacy_news && DISPLAY_SUMMARY_KINDS.contains(kind))
        .and_then(|_| object_field(&metadata, "summary"))
        .and_then(|summary| summary.get("classification"))
        .and_then(Value::as_str)
        .and_then(|value| match value {
            "to_read" => Some(ContentClassification::ToRead),
            "skip" => Some(ContentClassification::Skip),
            _ => None,
        });
    let primary_topic = (!is_legacy_news)
        .then(|| project_topics(&metadata, kind, version))
        .and_then(|topics| {
            topics
                .first()
                .and_then(|value| clean_optional_str(Some(value)))
        })
        .or_else(|| {
            (content.content_type == ContentType::News)
                .then(|| platform.clone())
                .flatten()
                .and_then(|value| clean_optional_str(Some(&value)))
        });
    let key_takeaway = extract_key_takeaway(&metadata);
    let metadata = sanitize_metadata_for_api(metadata);
    let top_comment = content_top_comment(
        platform.as_deref(),
        content.discussion_url.as_deref(),
        &metadata,
    );
    let comment_count = metadata.get("comment_count").and_then(Value::as_i64);
    let saved_source =
        saved_source_override.or_else(|| infer_saved_source(&metadata, row.is_saved_to_knowledge));
    let user_status = matches!(
        content.content_type,
        ContentType::Article | ContentType::Podcast
    )
    .then(|| "inbox".to_owned());
    Ok(ContentSummaryResponse {
        id: row.id,
        content_type: content.content_type,
        url: content.url,
        source_url: row.source_url.or(Some(row.url)),
        discussion_url: content.discussion_url,
        title: Some(content.display_title),
        source: row.source,
        platform,
        status: content.status,
        short_summary: content.summary_text.clone(),
        created_at: row.created_at,
        processed_at: row.processed_at,
        classification,
        publication_date: row.publication_date,
        is_read: row.is_read,
        is_saved_to_knowledge: row.is_saved_to_knowledge,
        knowledge_saved_at,
        news_article_url: content
            .news_fields
            .as_ref()
            .and_then(|fields| fields.article_url.clone()),
        news_discussion_url: content
            .news_fields
            .as_ref()
            .and_then(|fields| fields.discussion_url.clone()),
        news_key_points: content
            .news_fields
            .as_ref()
            .and_then(|fields| fields.key_points.clone()),
        news_summary: content.summary_text,
        user_status,
        image_url: content.image_url,
        thumbnail_url: content.thumbnail_url,
        primary_topic,
        top_comment,
        comment_count,
        feed_preview,
        artifact_type,
        preview_bullets,
        reason_to_read,
        key_takeaway,
        saved_source,
    })
}

fn infer_saved_source(metadata: &Map<String, Value>, is_saved: bool) -> Option<SavedSource> {
    if !is_saved {
        return None;
    }
    let is_x_bookmark = [
        ("submitted_via", "x_bookmarks"),
        ("tweet_snapshot_source", "x_bookmarks_sync"),
    ]
    .into_iter()
    .any(|(key, expected)| {
        string_field(metadata, key).is_some_and(|value| value.to_lowercase() == expected)
    });
    Some(if is_x_bookmark {
        SavedSource::XBookmark
    } else {
        SavedSource::Knowledge
    })
}
